### PR TITLE
Disallow overflow in v3d to v3s16 conversion

### LINKED ISSUE #8909: Bug: nodes/content-IDs can exist beyond world's edge (and be found by minetest.find_node_near and the like)
##### Minetest version

```
Minetest 5.1.0-dev-1d3f5af7 (Linux)
```

##### Summary
I've discovered that schematics can be used to place content-IDs outside the edge of the world. Nodes from these schematics, although not rendered on the client, nevertheless exist and can be found using `minetest.find_node_near`.

##### Steps to reproduce
1. Create a schematic and place it in such a way that it overlaps beyond the world edge.
2. Use `minetest.find_node_near` to search for a unique node in the schematic.
3. Observe that the returned coordinates lie outside the world edge.

Note: I originally discovered this bug while working with leaf-decay code, in an attempt improve its interaction with the edge of the world and `ignore` nodes. I was surprised to discover that the trunks in the sapling schematics were detectable, even when placed outside the world.

##### Solution
When placing a schematic near the world's edge, slice it.

### PR DESCRIPTION
**Pull request by HybridDog**
It affects minetest.get_node and other functions if the position is outside the map boundaries.
If I remember correctly, the emerge chat command could cause a huge map generation if it is used near map boundaries and one of the positions overflows.
I did not change `floatToInt` because numbers from lua are in double format.

The PR may fix #8909.

## To do

Test if there are any regressions

## How to test

You can install the [LuaCmd mod](https://wiki.minetest.net/Mods/LuaCmd) and execute this in the chat:
`/lua p = vector.round(me:get_pos()) p.x = p.x + 0x10000 print(minetest.pos_to_string(p) .. " " .. minetest.get_node(p).name)`
Without the PR, it shows the node at the player's feed although the position is not inside the map.

### CODE REVIEW

File: src/util/numeric.h [hunk 1]
```diff
@@ -282,10 +282,18 @@ inline v3s16 floatToInt(v3f p, f32 d)
...
 {
-	return v3s16(
+	// Adjust p for correct rounding
+	v3d p_adjusted(
 		(p.X + (p.X > 0 ? d / 2 : -d / 2)) / d,
 		(p.Y + (p.Y > 0 ? d / 2 : -d / 2)) / d,
 		(p.Z + (p.Z > 0 ? d / 2 : -d / 2)) / d);
+	// Do not overflow if p is outside map boundaries
+	const double s16_min = -32768.0;
+	const double s16_max = 32767.0;
```
Comment (SmallJoker) [inline src/util/numeric.h]: (double)S16_MIN
(double)S16_MAX

**Comment by paramat** [comment]
I doubt this fixes #8909 because https://github.com/minetest/minetest/issues/8909#issuecomment-533823601

**Comment by HybridDog** [comment]
> I doubt this fixes #8909 because #8909 (comment)

You could be right. Nonetheless, this PR could fix other issues mentioned above.

**Comment by paramat** [comment]
Hmm, i think this is the wrong place to do value checking, because it is a function that can be used very rapidly and you have added 3 'rangelims' to it, that are processed every time the function is run.
Value checking is more efficiently done elsewhere because in many situations the checks are then done only done once even if this function is then called many times.
:-1: 

So for 'emergeblocks', the value checking should be in the chat command.

**Comment by HybridDog** [comment]
> […] it is a function that can be used very often and very rapidly […]

This function is only used when vectors from lua are converted to v3s16 (as far as I can see).
I think  `minetest.bulk_set_node` is the only function which takes a lot of vectors from lua, and the rangelim overhead is probably very low.

**Comment by paramat** [comment]
> This function is only used when vectors from lua are converted to v3s16 (as far as I can see).

Currently that may be so, but it is highly likely to not be the case in the future.
The absolute overhead may be small, but proportional to the function it is very significant.
But anyway, all this is mostly irrelevant because it is a case of good practice to put the checks elsewhere so they are not unnecessarily repeated or can be left out if not critical or if there is no need for them.

**Comment by HybridDog** [comment]
> Currently that may be so, but it is highly likely to not be the case in the future.

I could rename the function do `doubleToIntClamped` or something similar so that it is obvious that it is possible to implement a slightly faster version.

> But anyway, all this is mostly irrelevant because it is a case of good practice to put the checks elsewhere so they are not unnecessarily repeated or can be left out if not critical or if there is no need for them.

I think the checks are always required when a mod passes a position to Minetest core.

**Comment by paramat** [comment]
> This function is only used when vectors from lua are converted to v3s16 (as far as I can see).

`inline v3s16 doubleToInt(v3d p, double d)` is used here:

https://github.com/minetest/minetest/blob/2384c10e104fd0c945f3677130e7d25c0a841482/src/script/common/c_converter.cpp#L286-L298

Those are used a lot in MT, for example in 'set node', 'get node' etc.
> If I remember correctly, the emerge chat command could cause a huge map generation if it is used near map boundaries and one of the positions overflows.

Then checks should be added to the emerge chat command, that is the better place for them for this problem, and they can then print a warning message.

Your change does a `rangelim` of each axis to the range of an s16, the resulting value is therefore a value equally as incorrect as an overflowing value, so there is no benefit. And a small performance decrease in heavily used Lua API functions.
:-1:

**Comment by HybridDog** [comment]
I agree; to solve the problem more cleanly, warning messages can be added to the chat commands and documentation about undefined behaviour for invalid positions can be added for the performance-critical functions.

**Pull request closed without merging**